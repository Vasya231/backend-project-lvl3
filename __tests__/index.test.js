import { promises as fs } from 'fs';
import path from 'path';
import os from 'os';
import nock from 'nock';

import downloadPage from '../src/index';

const downloadPageTest = (pageUrl, outputDir) => downloadPage(
  pageUrl,
  outputDir,
  { testMode: true },
);

const pathToInputData = (...pathParts) => path.join(__dirname, '__fixtures__', 'inputData', ...pathParts);
const pathToExpectedData = (...pathParts) => path.join(__dirname, '__fixtures__', 'expected', ...pathParts);

let tmpDir;
const expectedData = {};

const mockWorkingPage = (hostname, pageName = '') => {
  const pagePath = `/${pageName}`;
  nock(hostname)
    .get(pagePath)
    .replyWithFile(200, pathToInputData('page1.html'));
  nock(hostname)
    .get('/assets/img1.jpg')
    .replyWithFile(200, pathToInputData('page1_files', 'img1.jpg'));
  nock(hostname)
    .get('/assets/script1.js')
    .replyWithFile(200, pathToInputData('page1_files', 'script1.js'));
  nock(hostname)
    .get('/assets/styles.css')
    .replyWithFile(200, pathToInputData('page1_files', 'styles.css'));
};

beforeAll(async () => {
  mockWorkingPage('https://testhost.ru', 'page1');
  mockWorkingPage('https://testhostbaddir');
  mockWorkingPage('https://testhostnorights');
  mockWorkingPage('https://testhostfileexists');
  mockWorkingPage('https://testhostnorewrite');
  nock('https://testhost404')
    .get('/')
    .reply(404);
  nock('https://testhostnocode')
    .get('/')
    .replyWithError('!!!');
  expectedData.img1 = await fs.readFile(pathToExpectedData('testhost-ru-page1_files', 'assets-img1.jpg'));
  expectedData.script1 = await fs.readFile(pathToExpectedData('testhost-ru-page1_files', 'assets-script1.js'));
  expectedData.css1 = await fs.readFile(pathToExpectedData('testhost-ru-page1_files', 'assets-styles.css'));
});

beforeEach(async () => {
  tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'page-loader-'));
});

describe('Positive testing', () => {
  test('Should download page with local resources and save them under generated names in output folder', async () => {
    await downloadPageTest('https://testhost.ru/page1', tmpDir);
    const actualData = {};
    actualData.page1 = await fs.readFile(path.join(tmpDir, 'testhost-ru-page1.html'), 'utf-8');
    actualData.img1 = await fs.readFile(path.join(tmpDir, 'testhost-ru-page1_files', 'assets-img1.jpg'));
    actualData.script1 = await fs.readFile(path.join(tmpDir, 'testhost-ru-page1_files', 'assets-script1.js'));
    actualData.css1 = await fs.readFile(path.join(tmpDir, 'testhost-ru-page1_files', 'assets-styles.css'));
    expect(actualData.page1).toMatchSnapshot();
    expect(actualData.img1).toEqual(expectedData.img1);
    expect(actualData.script1).toEqual(expectedData.script1);
    expect(actualData.css1).toEqual(expectedData.css1);
  });
});

describe('Negative testing: invalid arguments', () => {
  test('Invalid directory name', () => {
    expect(() => downloadPageTest('https://testhost1.ru', {})).toThrow('Path to directory must be a string.');
  });

  test('Invalid page URL', () => {
    expect(() => downloadPageTest('xfdgrh', tmpDir)).toThrow('Invalid URL: xfdgrh');
  });
});

describe('Negative testing: file system errors', () => {
  test('Output directory doesn\'t exist', async () => {
    const pathToResources = path.join(tmpDir, 'whatever', 'testhostbaddir_files');
    const expectedErrorMessage = `Cannot create directory '${pathToResources}'. Reason: no such file or directory`;
    const nonexistantDirPath = path.join(tmpDir, '/whatever');
    await expect(downloadPageTest('https://testhostbaddir', nonexistantDirPath)).rejects.toThrow(expectedErrorMessage);
  });

  test('No permissions', async () => {
    const pathToResources = path.join(tmpDir, 'testhostnorights_files');
    const expectedErrorMessage = `Cannot create directory '${pathToResources}'. Reason: permission denied`;
    await fs.chmod(tmpDir, 0);
    await expect(downloadPageTest('https://testhostnorights', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });

  test('File already exists in place of directory', async () => {
    const existingFilePath = path.join(tmpDir, 'testhostfileexists_files');
    const expectedErrorMessage = `Cannot create directory '${existingFilePath}'. Reason: file already exists`;
    await fs.writeFile(existingFilePath, ' ');
    await expect(downloadPageTest('https://testhostfileexists', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });

  test('File already exists and cant be rewritten', async () => {
    const pathToPage = path.join(tmpDir, 'testhostnorewrite.html');
    const expectedErrorMessage = `Cannot open file '${pathToPage}'. Reason: permission denied`;
    await fs.writeFile(pathToPage, ' ');
    await fs.chmod(pathToPage, 0);
    await expect(downloadPageTest('https://testhostnorewrite', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });
});

describe('Negative testing: network errors', () => {
  test('Page does not exist', async () => {
    const expectedErrorMessage = 'Cannot load \'https://testhost404/\'. Reason: Request failed with status code 404';
    await expect(downloadPageTest('https://testhost404', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });

  test('No answer from host', async () => {
    const expectedErrorMessage = 'Cannot load \'https://169.254.0.1/\'. Reason: Timeout of 3000ms exceeded.';
    await expect(downloadPage('https://169.254.0.1', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });

  test('Host doesn\'t exist', async () => {
    const expectedErrorMessage = 'Cannot resolve hostname \'https://kjadfhkdjfh.dfjj/\'. Reason: DNS lookup failed';
    await expect(downloadPageTest('https://kjadfhkdjfh.dfjj', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });

  test('Server replied with an error', async () => {
    const expectedErrorMessage = 'Cannot load \'https://testhostnocode/\'. Reason: !!!';
    await expect(downloadPageTest('https://testhostnocode', tmpDir)).rejects.toThrow(expectedErrorMessage);
  });
});
