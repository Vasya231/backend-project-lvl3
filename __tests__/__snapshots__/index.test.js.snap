// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Positive testing Should download page with local resources and save them under generated names in output folder 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"utf8\\">
  <title>Test page 1</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
  <link rel=\\"stylesheet\\" href=\\"testhost-ru-page1_files/assets-styles.css\\">
</head>

<body>
  <div class=\\"container mt-5\\">
    <div class=\\"jumbotron\\">
      <h1>RSS reader</h1>
      <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
      <form class=\\"rss-form\\">
        <div class=\\"form-inline\\">
          <label class=\\"sr-only\\" for=\\"url-input\\">Input feed URL</label>
          <input type=\\"text\\" class=\\"form-control mr-sm-2 mb-2 mb-sm-0\\" required name=\\"url\\" placeholder=\\"rss-link\\" id=\\"url-input\\">
          <button type=\\"submit\\" class=\\"btn btn-primary\\" data-toggle=\\"tooltip\\" title>Add</button>
        </div>
        <div class=\\"feedback text-danger mt-1\\"></div>
      </form>
    </div>
    <script src=\\"testhost-ru-page1_files/assets-script1.js\\"></script>
    <img src=\\"testhost-ru-page1_files/assets-img1.jpg\\">
    <img src=\\"https://something.com/pic.jpg\\">

    <div class=\\"row\\">
      <div class=\\"col-12 col-md-5 mb-3\\">
        <div class=\\"rss-feeds list-group\\"></div>
      </div>
      <div class=\\"col-12 col-md-5 rss-items ml-2\\"></div>
    </div>
  </div>


</body>

</html>"
`;
